### ISSUE TITLE
JWT token doesn't work after update to 5.8.2

### ISSUE THREAD

**Issue by kurdyshev** [emqx/emqx#14330]
### What happened?
If there is a service that needs the ability to subscribe to system topics.
On versions 5.8.0 and 5.8.1 the service could authenticate using jwt token with pyload:
`{
  "mqtt": "y",
  "acl": {
    "all": [
      "#"
    ]
  }
}` 
But on version 5.8.2 you need to change the pyload to be able to authenticate:
`{
  "mqtt": "y",
  "acl": {
    "all": [
      "#"
    ],
    "sub": [
      "$SYS/#"
    ]
  }
}`

### What did you expect to happen?

We didn't expect such an update in the patch version.

### EMQX version

<details>

```console
$ ./bin/emqx_ctl broker
sysdescr  : EMQX
version   : 5.8.2
datetime  : 2024-12-03T23:29:01.543028433+00:00
uptime    : 3 days, 21 hours, 12 minutes, 34 seconds
```

</details>

**Comment by thalesmg** [comment]
Hi @kurdyshev .

I tested locally with 5.8.2 and could not reproduce the issue.  Using `"acl": { "all": [ "#" ] }` seems to work just fine in 5.8.2.

Can you provide more details about your configuration and logs, if any?

**Comment by alexander-eroma** [comment]
Hello @thalesmg, thanks for your reply.

I've included for you a more detailed description below.

For the JWT authentication method, EMQX 5.8.2 does not allow subscribing to any system topics (topics which start with `$SYS/`)  if a `$SYS` subscription is not explicitly permitted in the JWT ACL claim. That was not the case for EMQX 5.8.1, where `{ "all": [ "#" ] }` was enough to grant access to system topics subscription.

Please take a look at the full example below.

### EMQX 5.8.1

1. Run the EMQX as a Docker container:
```
docker run -d --name emqx -p 18083:18083 -p 1883:1883 emqx:5.8.1
```

2. Configure JWT authentication:
<img width="1191" alt="image" src="https://github.com/user-attachments/assets/b99031c2-e61c-49eb-994b-631f44de9c60">

3. Generate client JWT token with the following payload
```
{
  "acl": {
    "all": [
      "#"
    ]
  }
}
```
using the [jwt.io](https://jwt.io/)
```
eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhY2wiOnsiYWxsIjpbIiMiXX19.ovXSRjIs35vx3HLWcD30CdugkbQ-9dTJzhHndQQQrXQ
```

4. Connect the MQTT client using the JWT token (im using MQTTX GUI tool):
<img width="663" alt="image" src="https://github.com/user-attachments/assets/7f81889b-d54d-4b0d-998a-aea37351b397">

5. Subscribe to `$SYS/#` topic:
<img width="558" alt="image" src="https://github.com/user-attachments/assets/7263a596-8630-45ee-af54-232146398bdd">

Result:
Subscribed successfully:
<img width="682" alt="image" src="https://github.com/user-attachments/assets/c86a98f5-d590-4013-8a67-c31fc22d94a2">

### EMQX 5.8.2

Run the EMQX as a Docker container:
```
docker run -d --name emqx -p 18083:18083 -p 1883:1883 emqx:5.8.2
```

and perform steps 2-5 from the previous example.

Result:
The connection is terminated by the EMQX:
```
Failed to Subscribe $SYS/#, Error: Not authorized(Code: 135). Make sure the permissions are correct, and check MQTT broker ACL configuration
```

<img width="784" alt="image" src="https://github.com/user-attachments/assets/94e610d9-81eb-4bf3-a0f7-16ac1cc31f57">

Here is the corresponding EMQX log:
```
WARNING: Default (insecure) Erlang cookie is in use.
WARNING: Configure node.cookie in /opt/emqx/etc/emqx.conf or override from environment variable EMQX_NODE__COOKIE
WARNING: NOTE: Use the same cookie for all nodes in the cluster.
EMQX_RPC__PORT_DISCOVERY [rpc.port_discovery]: manual
EMQX_NODE__NAME [node.name]: emqx@172.17.0.2
Listener tcp:default on 0.0.0.0:1883 started.
Listener ssl:default on 0.0.0.0:8883 started.
Listener ws:default on 0.0.0.0:8083 started.
Listener wss:default on 0.0.0.0:8084 started.
Listener http:dashboard on :18083 started.
EMQX 5.8.2 is running now!
2024-12-04T22:03:32.903955+00:00 [warning] tag: AUTHZ, clientid: mqttx_2549f49e, msg: authorization_permission_denied, peername: 192.168.65.1:47493, username: test, topic: $SYS/#, action: SUBSCRIBE(Q1), source: jwt
```

To have the `$SYS/#` topics subscription with EMQX 5.8.2 we need to explicitly add the corresponding topic to the JWT ACL claim:
```
{
  "acl": {
    "all": [
      "#"
    ],
    "sub": [
       "$SYS/#"
    ]
  }
}
```

The new behaviour looks reasonable. However, it may be a breaking change for some users, so it is worth to be added into the release change log.

**Comment by thalesmg** [comment]
Hi @alexander-eroma , thanks for the extra details.

Indeed, according to the MQTT spec, `#` shouldn't match `$SYS/` topics:

> The Server MUST NOT match Topic Filters starting with a wildcard character (# or +) with Topic Names beginning with a $ character [MQTT-4.7.2-1]. 

So, if the previous behavior was that `$SYS/` could be subscribed to using only `#`, that was a bug that apparently got fixed in 5.8.2.  I checked and really in 5.8.1, under some circumstances that could happen, and in 5.8.2 no longer.  One [PR](https://github.com/emqx/emqx/pull/14004/) ended up fixing that bug.

So, this breaking change is actually a bug fix.
